compiler/src/main/scala/org/scalajs/core/compiler/PrepJSInterop.scala

```diff
@@ -539,90 +559,31 @@ abstract class PrepJSInterop extends plugins.PluginComponent
-              }
-            } else if (isJSGlobalScope(implDef)) {
-              JSNativeLoadSpec.Global(Nil)
-            } else if (sym.hasAnnotation(JSImportAnnotation)) {
-              val annot = sym.getAnnotation(JSImportAnnotation).get
-              val module = annot.stringArg(0).getOrElse {
-                "" // do not care because it does not compile anyway
-              }
-              annot.stringArg(1).fold {
-                JSNativeLoadSpec.Import(module, Nil)
-              } { pathName =>
-                val path = pathName.split('.').toList
-                JSNativeLoadSpec.Import(module, path)
-              }
-            } else {
-              val needsExplicitJSName = {
-                (enclosingOwner is OwnerKind.ScalaMod) &&
-                !sym.owner.isPackageObjectClass
-              }
-
-              if (needsExplicitJSName && !sym.hasAnnotation(JSNameAnnotation)) {
-                if (sym.isModuleClass) {
-                  reporter.error(implDef.pos,
-                      "Native JS objects inside non-native objects must " +
-                      "have an @JSName or @JSImport annotation")
-                } else {
-                  // This should be an error, but we erroneously allowed that before
-                  reporter.warning(implDef.pos,
-                      "Native JS classes inside non-native objects should " +
-                      "have an @JSName or @JSImport annotation. " +
-                      "This will be enforced in 1.0.")
-                }
-              }
-
-              val path = jsInterop.jsNameOf(sym).split('.').toList
-              JSNativeLoadSpec.Global(path)
-            }
-          }
-
+          val loadSpec = checkAndComputeJSNativeLoadSpecOf(implDef.pos, sym)
           jsInterop.storeJSNativeLoadSpec(sym, loadSpec)
 
           // Mark module classes as having the new format
           if (sym.isModuleClass)
             sym.addAnnotation(HasJSNativeLoadSpecAnnotation)
         } else {
           assert(sym.isTrait) // just tested in the previous `if`
-          if (sym.hasAnnotation(JSNameAnnotation)) {
-            reporter.warning(implDef.pos,
-                "Traits should not have an @JSName annotation, as it does " +
-                "not have any effect. This will be enforced in 1.0.")
-          }
-          if (sym.hasAnnotation(JSImportAnnotation)) {
-            reporter.error(implDef.pos,
-                "Traits may not have an @JSImport annotation")
+          for {
+            annot <- sym.annotations
+            annotSym = annot.symbol
+            if JSNativeLoadingSpecAnnots.contains(annotSym)
+          } {
+            if (annotSym == JSNameAnnotation) {
+              reporter.warning(implDef.pos,
+                  "Traits should not have an @JSName annotation, as it does " +
+                  "not have any effect. This will be enforced in 1.0.")
+            } else {
+              reporter.error(implDef.pos,
```

Comment:
Oh! Duh! It just clicked what you're saying! I thought you were saying *warn* instead of *error*. But you're talking about the *position* primarily. Yes you are absolutely right, it should be on the annotation. And that also explains what you were saying about the "two positions" ^^